src/bootstrap/cache.rs

```diff
@@ -208,18 +162,49 @@ impl<T: Hash + Clone + Eq> TyIntern<T> {
 pub struct Interner {
     strs: Mutex<TyIntern<String>>,
     paths: Mutex<TyIntern<PathBuf>>,
+    lists: Mutex<TyIntern<Vec<String>>>,
+}
+
+trait Internable: Clone + Eq + Hash + 'static {
+    fn intern_cache() -> &'static Mutex<TyIntern<Self>>;
+
+    fn intern(self) -> Interned<Self> {
+        Self::intern_cache().lock().unwrap().intern(self)
+    }
+}
+
+impl Internable for String {
+    fn intern_cache() -> &'static Mutex<TyIntern<Self>> {
+        &INTERNER.strs
+    }
+}
+
+impl Internable for PathBuf {
+    fn intern_cache() -> &'static Mutex<TyIntern<Self>> {
+        &INTERNER.paths
+    }
+}
+
+impl Internable for Vec<String> {
+    fn intern_cache() -> &'static Mutex<TyIntern<Self>> {
+        &INTERNER.lists
+    }
 }
 
 impl Interner {
     pub fn intern_str(&self, s: &str) -> Interned<String> {
         self.strs.lock().unwrap().intern_borrow(s)
     }
     pub fn intern_string(&self, s: String) -> Interned<String> {
-        self.strs.lock().unwrap().intern(s)
```

Comment:
This doesn't seem right; I would expect this code to remain the same, calling self.strs... or self.paths.. for the other methods, and the trait intern(self) would call INTERNER.intern_x rather than splitting out intern_cache and intern(self) methods.

Reply:
`intern_cache` is needed for the generic methods, e.g. https://github.com/rust-lang/rust/blob/d50583eaecdf88b5ee59e1d77fe3b4a8d87cbe6e/src/bootstrap/cache.rs#L112-L117
Without making that part of the trait, there's no way to be generic over the type, since the only available method is for interning, not retrieving.

I do agree this diff in particular is kind of strange, though - I've reverted it back to make it more clear rather than being maximally concise.

Reply:
If you like I can remove `intern_cache` and add a `Internable::get` method. But not sure that any of these are more clear than others; they all seem about the same to me.

Reply:
I'm happy with the diff in the latest form; I think there's possibly a refactor to the interning that we could do (maybe even just leaking everything...) but definitely doesn't need to block this.